### ISSUE TITLE
More form-global information

### ISSUE THREAD

**Issue by souldreamer** [Sawtaytoes/OneForm#11]
Now that I can use OneForm in my project (I set up a local npm registry to use the latest version of the code until you update it on npm), I ran into a small issue: the availability of information about the form without going through fields one by one. While developing the form, I find it useful to have a small component just print out the form state so I can see that things are wired correctly (especially while creating form components that might not communicate well with OneForm yet).

For the form data itself I got away with using a context that gets updated with the `onChange` function on the OneForm component (even has a nice debounce happening so that's good, but it would have been a bit more work if I wasn't using this pattern in other context providers and had a ready-made hook for making it). It seems like a pity to have the information in internal state and not expose it to users (who might need it for more complex use-cases, or just for debugging purposes).

For the errors, it seems way more complicated, and is an even more common use-case (showing a list of all the errors in one place; though it's probably bad UX, people still use it). You have a hook for it (`useErrorMessagesState`), but it's not exported out of the library, and even when it is, it flattens the error object, so we lose the field => errors object and only keep all the errors and functions to get the errors by field name.

**Comment by Sawtaytoes** [comment]
Look at `useFormSubmission`. I'm missing the ability to have a `getAllValues` function without calling `submitForm`. I don't want a subscriber though. I want it to be "call function, get state".

The error messages you can get aren't grouped by field name, That's something I also want to change.

**Comment by souldreamer** [comment]
Totally missed it in `useFormSubmission`. Think I was looking for a get all values there and got disappointed when I didn't find it :)).

Also, speaking of the `onChange` prop of `OneForm`, it's acting very strangely. Shortly after starting to interact with the form and stopping, the internal `onChange` state value seems to change to `flattenSegmentedFunctionObjectValues` (I'm using `why did you render?` and it gives me the info directly, a `useState` hook changed from my `onChange` function to the `flatten*` one.

**Comment by souldreamer** [comment]
Managed to isolate & reproduce. It happens when the `onChange` function changes twice.

Reproduction:
[https://codesandbox.io/s/oneform-onchange-breaking-xg502](https://codesandbox.io/s/oneform-onchange-breaking-xg502)

Steps:
- open up console so you can see the `console.log`s (they just print out the parameter `onChange` is called with
- interact with the form, and you'll see the logs
- press the "Try breaking" button twice (the `onChange` is wrapped in a `useCallback` that depends on state that the button sets)
- now interact with the form and the `onChange` won't be called anymore. From using WYDR on my own project, I'm guessing it somehow gets replaced with `flattenSegmentedFunctionObjectValues` (which would be a major possible bug, as IIRC it uses `useSubformState` and shares the mechanism with a lot of other things)

**Comment by Sawtaytoes** [comment]
Yeah, looks like something's wrong here. I verified this no-longer-firing-onChange bug even in `v1.8.0`.

When passing new `onChange` functions, it works fine once, but then never after.

---

The `flattenSegmentedFunctionObjectValues` is working properly. It first uses the root `onChange`. If any `useSubformEffect`s exist, then it creates a new `onChange` combining the outputs into a single `changedValues` object.

That's the intended behavior. This is still a bug, but the flatten function is correctly being called. I'm wondering if the root `onChange` is being lost though.

**Comment by Sawtaytoes** [comment]
I added a story to test this: 4c237472a00b7d8248f17683d582b37037596fb4

You can get to it here locally: http://localhost:6006/?path=/story/forms-oneform--updating-on-change

Looking into the issue now. Something interesting, the `count` in my example stays at `0` even after the first click when `onChange` still fires.

**Comment by Sawtaytoes** [comment]
This check denies functions (oops):
```js
      if (
        value
        && (
          typeof value
          !== 'function'
        )
      ) {
```

I know why it does, but I changed how that code works already yet left this in here.

Subforms are extremely powerful tools for OneForm (still officially undocumented as we test them out at work). We're exclusively using `useSubformEffect` instead of passing functions to the root OneForm. That's why this went unnoticed; it only affects functions passed to the root `OneForm`: `onChange` and `onSubmit`, and only if they're updated.

Fixed `v1.8.2`: b6f27d1efaea25322c9b2b41716ff8b6456a60fe
